contracts/community-sbt/src/lib.rs

```diff
@@ -138,21 +134,65 @@ impl Contract {
             }
             self.assert_ttl(ttl, max_ttl);
         }
+        if let Some(memo) = memo {
+            env::log_str(&format!("SBT renew memo: {}", memo));
+        }
+
+        let expires_at_ms = env::block_timestamp_ms() + ttl;
+        ext_registry::ext(self.registry.clone()).sbt_renew(tokens, expires_at_ms)
     }
 
     /// Revokes list of tokens. If `burn==true`, the tokens are burned (removed). Otherwise,
     /// the token expire_at is set to now, making the token expired. See `registry.sbt_revoke`
     /// for more details.
+    /// Only class minters are allowed to revoke tokens
     pub fn sbt_revoke(
         &mut self,
         tokens: Vec<TokenId>,
         burn: bool,
         memo: Option<String>,
     ) -> Promise {
-        self.assert_admin();
+        require!(!tokens.is_empty(), "tokens must be a non empty list");
+        let caller = env::predecessor_account_id();
+        ext_registry::ext(self.registry.clone())
+            .sbts(env::current_account_id(), tokens.clone())
+            .then(
+                Self::ext(env::current_account_id())
+                    .on_sbt_revoke_callback(&caller, tokens, burn, memo),
+            )
+    }
+
+    /// sbt_revoke callback. Checks if all the the tokens can be revoked by the caller
+    /// based on the return value from registry.sbts. If not panics with a error message
+    #[private]
+    pub fn on_sbt_revoke_callback(
+        &self,
+        caller: &AccountId,
+        tokens: Vec<TokenId>,
+        burn: bool,
+        memo: Option<String>,
+        #[callback_result] token_data: Result<Vec<Option<Token>>, near_sdk::PromiseError>,
+    ) -> Promise {
```

Comment:
I think the error is descriptive enough, there is no need for result. It shows the exact reason it failed. I tried the approach with the result but failed to succeed there are some traits implementations missing and the method becomes more complex since we need to add if statements in the callback

Reply:
Let me add it to the icebox